packages/vaex-core/vaex/expression.py

```diff
@@ -196,6 +197,140 @@ def __init__(self, expression):
         self.expression = expression
 
 
+class StructOperations(object):
```

Comment:
No need to inherit from object, what about inheriting from https://docs.python.org/3/library/collections.abc.html#collections.abc.Mapping ?

Reply:
I inherited from object to be consistent with the other operation classes (e.g. `StringOperations`). I agree, there shouldn't be a need to inherit from object because python 2 support for new style classes is probably not required anymore.

I really like the idea of inheriting from `collections.abc.Mapping` however, as you've pointed out, we get in trouble with duplicated field names here. Implementing `__getitem__` to handle duplicated fields should work just fine, but `__iter__` should return index positions (since they are truly unique) instead of labels which might be confusing when doing things like `for field, dtype in struct.items()` because field will be an integer instead of a string label. Perhaps we make things more complicated when trying to force the duplicated keys data type in the mapping interface. What do you think?

Reply:
I think we can support Mapping, but all methods should raise a KeyError when there are duplicate keys, and hint the user to use integers for keys. So in 90% of the cases, users can just use the Mapping interface.



> but `__iter__` should return index positions

I think it should return the keys (strings) but raise a KeyError in the case of duplicate keys.

The most robust code would then be (and used internally in Vaex):
```python
for i in range(len(df.x.struct)):
  name = df.x.struct.keys()[i]
  expression = df.x.struct[i]
  ...

# or
for i, name in enumerate(df.x.struct.keys()):
  expression = df.x.struct[i]
  ...

```

But you can often get away by doing:
```python
for name, expression in df.x.items():
  ...
```

@JovanVeljanoski and @mansenfranzen do you both agree with this?

We can take this a step further.. and also implement items to make it work with duplicate keys, the only thing that wouldn't work would be `__getitem__` with a string argument.

Reply:
Sounds reasonable to me. Here is a little prototype implementation which might help us to discuss it in more detail:

```python
import collections
import pyarrow as pa

values = [[1, 2, 3],
          [4, 5, 6],
          [7, 8, 9]]
names = ["col1", "col1", "col3"]
array = pa.StructArray.from_arrays(arrays=values, names=names)


class StructOperations(collections.abc.Mapping):
    def __init__(self, array):
        self.array = array
        
        keys = list(self.keys())
        self._duplicates = set([x for x in keys if keys.count(x) > 1])
               
    def __iter__(self):
        for name in self.keys():
            yield name
    
    def __getitem__(self, key):
        if key in self._duplicates:
            raise LookupError("Duplicated label: please use index position.")
        return self.array.field(key)
    
    def __len__(self):
        return len(self.array.type)
    
    def keys(self):
        return (field.name for field in self.array.type)
    
    def values(self):
        return (field.type for field in self.array.type)
    
    def items(self):
        return zip(self.keys(), self.values())
    
    def get(self, key):
        return self[key]
    

mapping = StructOperations(array)

mapping.get(0) # <pyarrow.lib.Int64Array object at 0x0000024AE5719408> [1, 2, 3]
mapping.get("col1") # LookupError
mapping.get("col3") # <pyarrow.lib.Int64Array object at 0x0000024AE5719D08> [4, 5, 6]

list(mapping.keys()) # ['col1', 'col1', 'col3']
list(mapping.values()) # [DataType(int64), DataType(string), DataType(int64)]
list(mapping.items()) # [('col1', DataType(int64)), ('col1', DataType(string)), ('col3', DataType(int64))]
```

I allowed string lookup values for `__getitem__`, too. The only way to run into a key error is indeed when using a string for duplicated fields. Otherwise, all methods work fine with duplicated field names. 

@maartenbreddels @JovanVeljanoski 
What do you think?

Reply:
Yes, I think this would be great to use (sorry for the late reply, vacation period :))